aws/rust-runtime/aws-config/src/json_credentials.rs

```diff
@@ -0,0 +1,304 @@
+use smithy_json::deserialize::token::skip_value;
+use smithy_json::deserialize::{json_token_iter, EscapeError, Token};
+use smithy_types::instant::Format;
+use smithy_types::Instant;
+use std::borrow::Cow;
+use std::error::Error;
+use std::fmt::{Display, Formatter};
+use std::time::SystemTime;
+
+#[derive(Debug)]
+pub(crate) enum InvalidJsonCredentials {
+    /// The response did not contain valid JSON
+    JsonError(Box<dyn Error + Send + Sync>),
+    /// The response was missing a required field
+    MissingField(&'static str),
+
+    /// Another unhandled error occured
+    Other(Cow<'static, str>),
+}
+
+impl From<EscapeError> for InvalidJsonCredentials {
+    fn from(err: EscapeError) -> Self {
+        InvalidJsonCredentials::JsonError(err.into())
+    }
+}
+
+impl From<smithy_json::deserialize::Error> for InvalidJsonCredentials {
+    fn from(err: smithy_json::deserialize::Error) -> Self {
+        InvalidJsonCredentials::JsonError(err.into())
+    }
+}
+
+impl Display for InvalidJsonCredentials {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            InvalidJsonCredentials::Other(msg) => write!(f, "{}", msg),
+            InvalidJsonCredentials::MissingField(field) => write!(
+                f,
+                "Expected field `{}` in IMDS response but it was missing",
+                field
+            ),
+            InvalidJsonCredentials::JsonError(json) => {
+                write!(f, "invalid JSON in IMDS response: {}", json)
+            }
```

Comment:
_nitpick_
Is it conventional to reverse the order these are defined in? I think when we write match statements for enums that the variants order should be the same as when we declared it. Just a style preference.

Reply:
oh yeah, I think that was just some copy paste evolution

Reply:
fixed